### ISSUE TITLE
Slow attribution possibly due to FeatureAttributionSequenceOutput.from_step_attributions

### ISSUE THREAD

**Issue by saxenarohit** [inseq-team/inseq#243]
Hi,

I have been trying the input_x_gradient method for attribution but it is running very slow.

I tried to time it and found this function call is taking around 57 seconds to run for context length  > 500. 
https://github.com/inseq-team/inseq/blob/de203611fdaf8dba5ed8bdffcd51d3b0e9b36ff9/inseq/data/attribution.py#L178

The attribution time is added on top of this. 
Is there a way to optimize this `from_step_attributions` function?

![Screenshot from 2024-01-04 17-57-49](https://github.com/inseq-team/inseq/assets/14234481/5ba0572c-54b4-4444-813e-2f3921dd6252)

Another function, get_scores_dict() also takes a lot of time to execute.
https://github.com/inseq-team/inseq/blob/de203611fdaf8dba5ed8bdffcd51d3b0e9b36ff9/inseq/data/attribution.py#L366

Thanks

**Comment by gsarti** [comment]
Hi @saxenarohit,

Thanks for reporting this! I'm trying to use IxG with a GPT-2 small for ~500 generation steps, but my execution times for `from_step_attributions` are much lower (roughly 0.5s). Could you provide me with the following information to help with the debugging?

- Is the model a decoder-only or an encoder-decoder?
- Are both source and target attributions computed, or just one of the two?
- Are you collecting step scores alongside attribution (i.e. do you specify any `step_scores` in `model.attribute`)?
- What is the size of the final source/target step score tensors produced by the attribution process?

You can answer the final question using the following code:

```python
import inseq

attribution_model = inseq.load_model("<YOUR_MODEL>", "input_x_gradient")

out = attribution_model.attribute(
    "<YOUR_INPUT>",
    # Customize with your parameters
    output_step_attributions=True,
)
if out.step_attributions[-1].source_attributions is not None:
	print(out.step_attributions[-1].source_attributions.shape)
if out.step_attributions[-1].target_attributions is not None:
	print(out.step_attributions[-1].target_attributions.shape)
```

Thank you in advance for your time!

**Comment by gsarti** [comment]
Thank you! I'm profiling the code and the culprit appears to be the `get_sequences_from_batched_steps` method in `torch_utils`.

```python
import torch
import inseq
from inseq.data import FeatureAttributionSequenceOutput

import cProfile

inseq_model = inseq.load_model(
	"princeton-nlp/Sheared-LLaMA-1.3B",
	"input_x_gradient",
	model_kwargs={"torch_dtype": torch.float16}
)
out = inseq_model.attribute(
    input_texts="This is a very long text"*50,
    output_step_attributions=True,
	generation_args={"do_sample": False, "max_length": 512}
)

profiler = cProfile.Profile()
profiler.runcall(FeatureAttributionSequenceOutput.from_step_attributions,
    attributions=out.step_attributions,
    pad_id=inseq_model.pad_token,
    has_bos_token=False,
)
profiler.print_stats("tottime")
```

Result:

```bash
         1116 function calls in 57.366 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      211   23.722    0.112   23.722    0.112 {built-in method torch.cat}
        1   11.703   11.703   11.703   11.703 {built-in method torch.stack}
      211   10.977    0.052   10.977    0.052 {built-in method torch.ones}
        1   10.867   10.867   57.275   57.275 torch_utils.py:176(get_sequences_from_batched_steps)
        1    0.089    0.089   57.365   57.365 attribution.py:177(from_step_attributions)
      216    0.004    0.000    0.004    0.000 {built-in method builtins.max}
      212    0.001    0.000    0.001    0.000 attribution.py:199(<genexpr>)
        3    0.001    0.000    0.001    0.000 torch_utils.py:183(<listcomp>)
```

In practice, the `get_sequences_from_batched_steps` operates naively on the batched steps, without exploiting their regularity, namely:
- When `batch_size` shrinks when moving forward with the generation, it is an indicator that some of the sentences were terminated in the previous batched step -> we can use the sequence length from that step to perform the padding, instead of looping on the tensor.
- Every step is exactly 1 token longer than the previous.
- In the edge case where a single sequence is provided, no loop is needed and the whole sequence can simply be stacked.

The fix is WIP and will be applied in #245!

**Comment by gsarti** [comment]
@saxenarohit using #245 I see now ~0.2s for `from_step_attributions` attributions in the scenario described above. Could you try to check out the branch to confirm the fix is working for you? Thanks in advance!

**Comment by gsarti** [comment]
I am using a VM with 64 CPUs, 32GB of RAM and 1x RTX 4090. This is my code, copied from the one above:

```python
%pip install git+https://github.com/inseq-team/inseq.git@speed-mem-opt

import torch
import inseq
from inseq.data import FeatureAttributionSequenceOutput

import cProfile

inseq_model = inseq.load_model(
	"princeton-nlp/Sheared-LLaMA-1.3B",
	"input_x_gradient",
	model_kwargs={"torch_dtype": torch.float16}
)
out = inseq_model.attribute(
    input_texts="This is a very long text"*50,
    output_step_attributions=True,
	generation_args={"do_sample": False, "max_length": 512}
)

profiler = cProfile.Profile()
profiler.runcall(FeatureAttributionSequenceOutput.from_step_attributions,
    attributions=out.step_attributions,
    pad_id=inseq_model.pad_token,
    has_bos_token=False,
)
profiler.print_stats("tottime")
```

And these are the profiler stats:

```bash
         687 function calls in 0.269 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      210    0.156    0.001    0.156    0.001 {built-in method torch.nn.functional.pad}
        1    0.111    0.111    0.111    0.111 {built-in method torch.stack}
        1    0.001    0.001    0.269    0.269 torch_utils.py:188(get_sequences_from_batched_steps)
        1    0.000    0.000    0.269    0.269 attribution.py:177(from_step_attributions)
      212    0.000    0.000    0.000    0.000 attribution.py:199(<genexpr>)
        1    0.000    0.000    0.000    0.000 attribution.py:206(<listcomp>)
        1    0.000    0.000    0.000    0.000 torch_utils.py:212(<listcomp>)
        1    0.000    0.000    0.000    0.000 _tensor.py:849(split)
        1    0.000    0.000    0.000    0.000 attribution.py:241(<listcomp>)
        1    0.000    0.000    0.000    0.000 {built-in method torch.split}
        2    0.000    0.000    0.000    0.000 misc.py:174(drop_padding)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.all}
        1    0.000    0.000    0.000    0.000 {method 'squeeze' of 'torch._C._TensorBase' objects}
      219    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        1    0.000    0.000    0.000    0.000 torch_utils.py:222(<listcomp>)
        4    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
        3    0.000    0.000    0.000    0.000 {built-in method builtins.max}
        1    0.000    0.000    0.000    0.000 _VF.py:26(__getattr__)
        1    0.000    0.000    0.000    0.000 attribution.py:716(__post_init__)
        1    0.000    0.000    0.000    0.000 attribution.py:150(__post_init__)
        1    0.000    0.000    0.000    0.000 attribution.py:205(<listcomp>)
        2    0.000    0.000    0.000    0.000 typing.py:20(__eq__)
        1    0.000    0.000    0.000    0.000 <string>:2(__init__)
        2    0.000    0.000    0.000    0.000 torch_utils.py:212(<genexpr>)
        1    0.000    0.000    0.000    0.000 attribution.py:421(get_sequence_cls)
        1    0.000    0.000    0.000    0.000 {built-in method torch._C._has_torch_function_unary}
        1    0.000    0.000    0.000    0.000 attribution.py:217(<listcomp>)
        2    0.000    0.000    0.000    0.000 misc.py:177(<lambda>)
        2    0.000    0.000    0.000    0.000 {method 'keys' of 'dict' objects}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.getattr}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 attribution.py:212(<listcomp>)
        1    0.000    0.000    0.000    0.000 attribution.py:256(<listcomp>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.min}
        2    0.000    0.000    0.000    0.000 attribution.py:216(<genexpr>)
        1    0.000    0.000    0.000    0.000 attribution.py:266(<listcomp>)
        1    0.000    0.000    0.000    0.000 {method 'update' of 'dict' objects}
        1    0.000    0.000    0.000    0.000 {method 'append' of 'list' objects}
```
